File: Pedagogie/Didactica Informaticii/problema.py
# Citirea datelor
n = int(input("Introduceți numărul de spectacole: "))
spectacole = []


for i in range(n):
# Citim o linie, o despărțim după spații și convertim în întregi
    date = input().split()
    start = int(date[0])
    final = int(date[1])
    # Adăugăm un tuplu de forma (ID, start, final)
    spectacole.append((i + 1, start, final))


# Căutăm la fiecare pas minimul din porțiunea nesortată a listei
for i in range(len(spectacole) - 1):
    min_idx = i
    for j in range(i + 1, len(spectacole)):
        # Comparăm după ora de final (indexul 2 din tuplu: (id, start, final))
        if spectacole[j][2] < spectacole[min_idx][2]:
            min_idx = j
           
    # Dacă am găsit un spectacol care se termină mai devreme, interschimbăm (swap)
    if min_idx != i:
        spectacole[i], spectacole[min_idx] = spectacole[min_idx], spectacole[i]




# Pasul Greedy: Alegem primul spectacol sortat
alese = [spectacole[0]]


# Verificăm restul spectacolelor
for i in range(1, len(spectacole)):
    # spectacole[i][1] este ora de start a spectacolului analizat
    # alese[-1][2] este ora de final a ultimului spectacol adăugat
    if spectacole[i][1] >= alese[-1][2]:
        alese.append(spectacole[i])


# Afișarea datelor
print("Spectacolele alese sunt:", end=" ")
for sp in alese:
    print(sp[0], end=" ")


print(f"\nNumărul de spectacole este {len(alese)}") # Se putea face și cum am făcut la C++, dar am ales să folosesc proprietățiile limbajului

File: Pedagogie/Didactica Informaticii/problema.cpp
#include <iostream>
using namespace std;


// Structura pentru a grupa datele unui spectacol
struct Spectacol {
    int id;
    int start; 
    int final; };


int main() {
    int n;
    Spectacol S[100]; 
    cin >> n;
    for (int i = 0; i < n; i++) {
        S[i].id = i + 1; // Salvăm numărul spectacolului (1, 2, 3...)
        cin >> S[i].start >> S[i].final;
    }


    // Sortare crescătoare după ora de final (Metoda Bulelor)
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (S[j].final > S[j+1].final) {
                // Interschimbăm structurile complet
                Spectacol aux = S[j];
                S[j] = S[j+1];
                S[j+1] = aux;
            }
        }
    }


    // Pasul Greedy: Alegem primul spectacol (cel care se termină cel mai repede)
    cout << "Spectacolele alese sunt ";
    int ultimul_f = S[0].final;
    cout << S[0].id << " ";
    int count = 1;


    // Verificăm restul spectacolelor
    for (int i = 1; i < n; i++) {
        // Dacă spectacolul curent începe după sau exact când se termină ultimul ales
        if (S[i].start >= ultimul_f) {
            cout << S[i].id << " ";
            ultimul_f = S[i].final; // Actualizăm ora de final curentă
            count++;
        }
    }


    cout << "\nNumarul de spectacole este " << count << "\n";


    return 0;
}
